probe-rs/src/debug/registers.rs

```diff
@@ -1,162 +1,243 @@
-            register_description: register_file,
-            values: HashMap::new(),
-            architecture: core.architecture(),
-            address_size: pc_desc.size_in_bytes(),
-        };
-
-        for i in 0..num_platform_registers {
-            let result: Result<u64, crate::Error> =
-                core.read_core_reg(register_file.platform_register(i));
-            match result {
-                Ok(value) => registers.values.insert(i as u32, value),
-                Err(e) => {
-                    log::warn!("Failed to read value for register {}: {}", i, e);
-                    None
+        let mut debug_registers = Vec::<DebugRegister>::new();
+
+        let all_registers = [
+            (RegisterGroup::Base, register_file.platform_registers),
+            (RegisterGroup::Argument, register_file.argument_registers),
+            (RegisterGroup::Result, register_file.result_registers),
+            (
+                RegisterGroup::Singleton,
+                &[register_file.frame_pointer.to_owned()],
+            ),
+            (
+                RegisterGroup::Singleton,
+                &[register_file.program_counter.to_owned()],
+            ),
+            (
+                RegisterGroup::Singleton,
+                &[register_file.return_address.to_owned()],
+            ),
+            (
+                RegisterGroup::Singleton,
+                &[register_file.stack_pointer.to_owned()],
+            ),
+        ];
+
+        for (register_group, register_group_members) in all_registers {
+            for (dwarf_id, platform_register) in register_group_members.iter().enumerate() {
+                // Check to ensure the register type is compatible with u64.
+                if matches!(
+                    platform_register.data_type(),
+                    RegisterDataType::UnsignedInteger
+                ) && platform_register.size_in_bits() <= 64
+                {
+                    if let Some(special_register) = debug_registers
+                        .iter_mut()
+                        .find(|debug_register| debug_register.id == platform_register.id)
+                    {
+                        // Some register definitions are descriptive for registers defined with the same [`RegisterId`] elsewhere, so we treat them differently.
+                        special_register.special_name = Some(platform_register.name);
+                    } else {
+                        // It is safe for us to push a new [`DebugRegister`]
+                        debug_registers.push(DebugRegister {
+                            register_file,
+                            group: register_group,
+                            name: platform_register.name(),
+                            special_name: None,
+                            id: platform_register.id,
+                            dwarf_id: if matches!(register_group, RegisterGroup::Base) {
+                                Some(dwarf_id as u16)
```

Comment:
This is probably true for most regs, but I don't think it is for all right now.  I think this is another argument to putting dwarf_id on `RegisterDescription` so we can avoid assumptions like this.

Reply:
Good catch. I've added a TODO to that effect.